src/iris/api.py

```diff
@@ -4036,36 +4056,57 @@ class Applications(object):
-
-            cursor.execute(get_supported_application_modes_query, app['id'])
-            app['supported_modes'] = [row['name'] for row in cursor]
+        query_limit = req.get_param_as_int('limit')
+        req.params.pop('limit', None)
+        requested_fields = req.get_param_as_list('fields')
+        fields = [f for f in requested_fields if f in application_columns] if requested_fields else None
+        if not fields:
+            fields = application_columns
+        req.params.pop('fields', None)
 
-            cursor.execute(get_application_owners_query, app['id'])
-            app['owners'] = [row['name'] for row in cursor]
+        query = application_query % ', '.join(application_columns[f] for f in fields)
 
-            cursor.execute(get_application_custom_sender_addresses, app['id'])
-            app['custom_sender_addresses'] = {row['mode_name']: row['address'] for row in cursor}
+        where = ['`auth_only` is False']
+        where += gen_where_filter_clause(connection, application_filters, application_filter_types, req.params)
 
-            cursor.execute(get_application_categories, app['id'])
-            app['categories'] = [row for row in cursor]
+        if where:
+            query = query + ' WHERE ' + ' AND '.join(where)
+        query += ' ORDER BY `application`.`name` ASC'
+        if query_limit is not None:
+            query += ' LIMIT %s' % query_limit
 
-            del app['id']
+        cursor.execute(query)
+        apps = cursor.fetchall()
+        if requested_fields is None:
+            # TODO: make these values individually selectable via fields
+            for app in apps:
+                app['title_variable'] = None
+                cursor.execute(get_vars_query, app['id'])
+                app['variables'] = []
+                app['required_variables'] = []
+                for row in cursor:
+                    app['variables'].append(row['name'])
+                    if row['required']:
+                        app['required_variables'].append(row['name'])
+                    if row['title_variable'] == 1:
+                        app['title_variable'] = row['name']
+
+                cursor.execute(get_default_application_modes_query, app['name'])
+                app['default_modes'] = {row['priority']: row['mode'] for row in cursor}
+
+                cursor.execute(get_supported_application_modes_query, app['id'])
+                app['supported_modes'] = [row['name'] for row in cursor]
+
+                cursor.execute(get_application_owners_query, app['id'])
+                app['owners'] = [row['name'] for row in cursor]
+
+                cursor.execute(get_application_custom_sender_addresses, app['id'])
+                app['custom_sender_addresses'] = {row['mode_name']: row['address'] for row in cursor}
+
+                cursor.execute(get_application_categories, app['id'])
+                app['categories'] = [row for row in cursor]
```

Comment:
This looks like it will run these 5 queries for every application returned by the above the query serially, if requested_fields is not specified as a url param? 

As we have lots of applications configured, and as afaik this endpoint is hit frequently to populate things like app names in typeaheads/etc, I'm concerned this will make things noticeably slow.

Is it possible to instead do something like:

- After we run `apps = cursor.fetchall()`, go through and get every ID returned 
- afterwards, run all of these get_application_*_query queries but using "where app id in LISTOFIDS" 
- go through those result sets and interpolate them into the apps list we got back on line 4079?

This way we will only be running 6 queries total instead of 1 + (5*268)

Reply:
This was already being run every time for this endpoint, I'm guessing since originally we had very few apps whoever wrote it was not very concerned about performance. I was originally planning to optimize it when I made these additional fields selectable once I would be changing this code then. I had some time so I went ahead and implemented that optimization as well as the TODO of making these other fields individually selectable